blueocean-core-js/src/js/UrlBuilder.ts

```diff
@@ -0,0 +1,302 @@
+    organizationName: string,
+    pipelineFullName: string, // Includes folder path
+    detailName: string, // Either the branchName or the pipeline short name if not multibranch
+    runId: string,
+}
+
+/**
+ * Parses the REST link to a pipeline run, extracts and unencodes the separate parts
+ *
+ * @param restUrl
+ * @returns {{organizationName: string; fullName: string; detailName: string; runId: string}}
+ */
+function parseRestRunUrl(restUrl: string): ResourceIdentifiers {
+
+    // FIXME: before exporting, make this more flexible so it can be a complete compliment to buildRestUrl
+
+    const tokens = trimEmptyTokens(restUrl.split('/'));
+
+    // given the following URL '/blue/rest/organizations/jenkins/pipelines/folder1/pipelines/folder2/pipelines/folder3/pipelines/jdl-2
+    // /branches/experiment%252Fbuild-locally-docker/runs/21/
+
+    const organizationName = decodeURIComponent(tokens[3]);
+    const isMultiBranch = tokens[tokens.length - 4] === 'branches';
+
+    const fullNameStart = 4;
+    const fullNameEnd = isMultiBranch ? tokens.length - 4 : tokens.length - 2;
+    // grab the tokens that make up the full name, then filter out the even values ('/pipelines')
+    // so the clean folder path is returned, e.g. f1/pipelines/f2/pipelines/f3/pipelines/foo => f1/f2/f3/foo
+    const pipelineFullName = tokens
+        .slice(fullNameStart, fullNameEnd)
+        .filter((name, index) => index % 2 === 1)
+        .join('/');
+
+    const pipelineName = tokens[fullNameEnd - 1];
+    const branchName = isMultiBranch ? tokens[tokens.length - 3] : '';
+    const runId = tokens[tokens.length - 1];
+
+    const detailName = decodeURIComponent(isMultiBranch ? decodeURIComponent(branchName) : pipelineName);
+
+    // fail fast
+    if (!organizationName || !pipelineFullName || !detailName || !runId) {
+        throw new Error('Could not extract URI components');
+    }
+
+    return {
+        organizationName,
+        pipelineFullName,
+        detailName,
+        runId,
+    };
+}
+
+type RunDetailsWithSelfLink = {
+    // FIXME: We need a canonical typedef for run details object
+    _links: { self: { href: string } }
+};
+
+/**
+ * Builds a run details view url from a RunDetails object
+ */
+export function buildRunUrlForDetails(runDetails: RunDetailsWithSelfLink) {
+    const restUrl = runDetails._links.self.href;
+    return buildRunUrlForRestUrl(restUrl);
+}
+
+/**
+ * Builds a run details view url from a run's REST link URL
+ */
+export function buildRunUrlForRestUrl(restUrl: string) {
+    const {organizationName, pipelineFullName, detailName, runId} = parseRestRunUrl(restUrl);
+
+    return buildRunUrl(organizationName, pipelineFullName, detailName, runId);
+}
+
+/**
+ * Builds a run details view url from identifiers
+ */
+export function buildRunUrl(organizationName, pipelineFullName, branchOrPipelineName, runId, tabName = 'pipeline') {
+    const baseUrl =
+        `/organizations/${encodeURIComponent(organizationName)}` +
```

Comment:
NIT: either `+` or backtick style, not both

Reply:
It's just too long to read all in one, no need to break up the individual "section" templates IMO